client/app/dbaas/logs/detail/streams/add/logs-streams-add.controller.js

```diff
@@ -1,7 +1,84 @@
 class LogsStreamsAddCtrl {
-    constructor ($stateParams) {
+    constructor ($q, $state, $stateParams, LogsStreamsService, ControllerHelper, CloudMessage) {
+        this.$q = $q;
+        this.$state = $state;
         this.$stateParams = $stateParams;
         this.serviceName = this.$stateParams.serviceName;
+        this.LogsStreamsService = LogsStreamsService;
+        this.ControllerHelper = ControllerHelper;
+        this.CloudMessage = CloudMessage;
+        this.isEdit = false;
+        this.compressionAlgorithms = this.LogsStreamsService.getCompressionAlgorithms();
+        this.storageDurations = this.LogsStreamsService.getStorageDurations();
+
+        this.initLoaders();
+    }
+
+    /**
+     * initializes options list
+     *
+     * @memberof LogsStreamsHomeCtrl
+     */
+    initLoaders () {
+        this.options = this.ControllerHelper.request.getArrayLoader({
+            loaderFunction: () => this.LogsStreamsService.getSubscribedOptions(this.serviceName)
+        });
+        this.options.load();
+
+        if (this.$stateParams.streamId) {
+            this.isEdit = true;
+            this.stream = this.ControllerHelper.request.getHashLoader({
+                loaderFunction: () => this.LogsStreamsService.getStream(this.serviceName, this.$stateParams.streamId)
+            });
+            this.stream.load();
+        } else {
+            this.isEdit = false;
+            this.stream = {
+                data: {
+                    coldStorageCompression: this.compressionAlgorithms[0].value,
+                    coldStorageRetention: this.storageDurations[0].value,
+                    coldStorageNotifyEnabled: true,
+                    coldStorageEnabled: false,
+                    webSocketEnabled: true
+                }
+            };
+        }
+    }
+
+    /**
+     * update stream
+     *
+     * @memberof LogsStreamsHomeCtrl
+     */
+    updateStream () {
+        if (this.form.$invalid) {
+            return this.$q.reject();
+        }
+        this.CloudMessage.flushChildMessage();
+        this.saving = this.ControllerHelper.request.getHashLoader({
+            loaderFunction: () =>
+                this.LogsStreamsService.updateStream(this.$stateParams.serviceName, this.stream.data)
+                    .then(() => this.$state.go("dbaas.logs.detail.streams"))
```

Comment:
You should use NavigationService.previousState.go() or $window.history.back()

Reply:
Back button must always take user back to streams home page. So I can not use NavigationService.previousState.go() or $window.history.back().

When user visits edit stream page from bookmark url, using $window.history.back() will break

Reply:
Back button must always take user back to streams home page. So I can not use NavigationService.previousState.go() or $window.history.back(). => I guess this depends on your opinion.  I'd be curious to know what is the intended behavior if => 
From my gridview page I click add, I add something. => I expect to go back to the gridview
From an unrelated page I click add something, I add something => I expect to go back to where?  I'd rather go back to where I was than to go back to the gridview

When user visits edit stream page from bookmark url, using $window.history.back() will break => CloudNavigation take care of that for you :).  If there is nothing in the history, we send you back to the current active tab's state.

Reply:
cool... I did not know that CloudNavigation takes care of it. I was thinking that it will navigate user back to login page because the user could have entered this page by entering is credentials. i will change it